kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java

```diff
@@ -18,26 +20,57 @@
  * Sends a single request frame, waits for a response then closes the channel
  */
 public class KafkaClientHandler extends ChannelInboundHandlerAdapter {
-    private final DecodedRequestFrame<?> decodedRequestFrame;
-    private final CompletableFuture<DecodedResponseFrame<?>> onResponse = new CompletableFuture<>();
+    private final List<DecodedRequestFrame<?>> decodedRequestFrames;
+    private final DecodedRequestFrame<?> initialRequest;
+    List<DecodedResponseFrame<?>> responses = new ArrayList<>();
+    private final CompletableFuture<List<DecodedResponseFrame<?>>> onResponse = new CompletableFuture<>();
+    private final CompletableFuture<Void> initialResponseReceived = new CompletableFuture<>();
 
     /**
      * Creates a KafkaClientHandler
-     * @param decodedRequestFrame the single request to send
+     * @param decodedRequestFrames the requests to send
      */
-    public KafkaClientHandler(DecodedRequestFrame<?> decodedRequestFrame) {
-        this.decodedRequestFrame = decodedRequestFrame;
+    public KafkaClientHandler(DecodedRequestFrame<?> initialRequest, List<DecodedRequestFrame<?>> decodedRequestFrames) {
+        this.decodedRequestFrames = decodedRequestFrames;
+        this.initialRequest = initialRequest;
+        if (this.initialRequest == null) {
+            initialResponseReceived.complete(null);
+        }
     }
 
     @Override
     public void channelActive(ChannelHandlerContext ctx) {
-        ctx.writeAndFlush(decodedRequestFrame);
+        if (initialRequest != null) {
+            ctx.writeAndFlush(initialRequest);
+        }
+        initialResponseReceived.thenAccept(unused -> {
+            writeAllMessages(ctx);
+        });
+    }
+
+    private void writeAllMessages(ChannelHandlerContext ctx) {
+        for (DecodedRequestFrame<?> decodedRequestFrame : decodedRequestFrames) {
+            ctx.write(decodedRequestFrame);
```

Comment:
IIRC calling  ctx.write from a thread that is not a netty event loop thread will schedule the work on the netty thread.  Wondering if you'd be better scheduling the loop and the flush on the event loop.

Reply:
The future was completed on the netty thread, but I think it'll be clearer using a boolean field instead and calling writeAllMessages in the `channelRead`.